File: Data Structures & Algorithms/quickSort/submission-5.java
// Definition for a pair.
// class Pair {
//     int key;
//     String value;
//
//     public Pair(int key, String value) {
//         this.key = key;
//         this.value = value;
//     }
// }
class Solution {
    public List<Pair> quickSort(List<Pair> pairs) {
        quickSort(pairs, 0, pairs.size() - 1);
        return pairs;
    }

    public void quickSort(List<Pair> arr, int s, int e) {
        if (e - s + 1 <= 1) return;
        Pair pivot = arr.get(e);
        int left = s;

        for (int i = s; i < e; i ++) {
            if (arr.get(i).key < pivot.key) {
                Pair temp = arr.get(left);
                arr.set(left, arr.get(i));
                arr.set(i, temp);
                left++;
            }
        }

        arr.set(e, arr.get(left));
        arr.set(left, pivot);

        quickSort(arr, s, left - 1);
        quickSort(arr, left + 1, e);
    }
}

File: Data Structures & Algorithms/mergeSort/submission-5.cpp
class Solution {
public:
    vector<Pair> mergeSort(vector<Pair>& pairs) {
        mergeSortHelper(pairs, 0, pairs.size() - 1);
        return pairs;
    }

    void mergeSortHelper(vector<Pair>& pairs, int s, int e) {
        if (e - s + 1 <= 1) return;
        
        int mid = (s + e) / 2;

        mergeSortHelper(pairs, s, mid);
        mergeSortHelper(pairs, mid + 1, e);
        merge(pairs, s, mid, e);
    }

    void merge(vector<Pair>& arr, int s, int mid, int e) {
        vector<Pair> L(arr.begin() + s, arr.begin() + mid + 1);
        vector<Pair> R(arr.begin() + mid + 1, arr.begin() + e + 1);

        int i = 0;
        int j = 0;
        int k = s;

        while (i < L.size() && j < R.size()) {
            if (L[i].key <= R[j].key) {
                arr[k++] = L[i++];
            } else {
                arr[k++] = R[j++];
            }
        }

        while (i < L.size()) {
            arr[k++] = L[i++];
        }

        while (j < R.size()) {
            arr[k++] = R[j++];
        }
    }
};
